Send Describe ACL requests to the controller

graphcareful opened the pull request:
Send Describle ACL requests to the controller node so clients do not observe stale state

Fixes: https://github.com/redpanda-data/redpanda/issues/9266

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

### Improvements

* Describe ACL requests will be forwarded to the controller node so that the most up-to-date state is available.

BenPope commented on src/v/cluster/topics_frontend.h:
```diff
@@ -24,6 +24,7 @@
 #include "model/record.h"
 #include "model/timeout_clock.h"
 #include "partition_balancer_types.h"
+#include "security/fwd.h"
```
Thank you!

BenPope commented on src/v/cluster/service.cc:
```diff
@@ -777,4 +777,23 @@ service::do_get_partition_state(partition_state_request req) {
+             })
+      .then([](result<std::vector<acl_resource>> resources) mutable {
+          if (resources.has_error()) {
+              return describe_acls_reply{
+                .error_code = errc(resources.error().value())};
```
This conversion feels a bit dirty. Is it guaranteed that `resources.error().category()` is always `cluster::errc_category`? I always feel like a `std::errc` of some sort could pop out of the network call.

NyaliaLui replied:
Or what if we make the error code in the reply a `std::errc` type?

BenPope replied:
> Or what if we make the error code in the reply a `std::errc` type?

It ends up going through a `std::error_code` elsewhere.

BenPope commented on src/v/kafka/server/handlers/describe_acls.cc:
```diff
@@ -24,55 +26,53 @@
+    explicit describe_acls_exception(std::error_code ec)
+      : _ec(ec) {}
+
+    cluster::errc error_code() const { return cluster::errc(_ec.value()); }
```
This conversion is unsafe, too. It might be simpler to avoid this exception altogether - it's only used in a single place to skip over a single line of code. You're still left with converting a `std::error_code` to a `cluster::errc` somewhere, though.

graphcareful replied:
Fixed

NyaliaLui commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -1512,4 +1515,86 @@ topics_frontend::get_partition_state(model::ntp ntp) {
+    }
+
+    std::vector<acl_resource> resources;
+    for (const auto& entry : entries) {
+        acl_resource resource;
+
+        // resource pattern
+        resource.type = entry.first.resource();
+        resource.name = entry.first.name();
+        resource.pattern_type = entry.first.pattern();
+
+        // acl entries
+        for (auto& acl : entry.second) {
+            // ignore ephemeral_users
+            auto ephemeral_user = security::principal_type::ephemeral_user;
+            if (acl.principal().type() == ephemeral_user) {
+                continue;
+            }
+            acl_description desc{
+              .principal = acl.principal(),
+              .host = acl.host(),
+              .operation = acl.operation(),
+              .permission_type = acl.permission(),
+            };
+            resource.acls.push_back(std::move(desc));
+        }
```
OUt of curiosity, do we skip ephemeral users because they cannot have ACLs or something?

BenPope replied:
> OUt of curiosity, do we skip ephemeral users because they cannot have ACLs or something?

They can have ACLs, but they shouldn't be returned through the Kafka API.

dotnwat commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -1512,4 +1515,86 @@ topics_frontend::get_partition_state(model::ntp ntp) {
 }
 
+std::vector<acl_resource>
+topics_frontend::do_describe_acls(const security::acl_binding_filter& filter) {
```
good spot same question

graphcareful replied:
Nice, i wasn't familiar with that frontend, that makes sense to me too

dotnwat reviewed (commented):
was there any traction with the comment I posted on team-replication about a generic solution using a query for the controller term? not that we need to do that now that this is written, but just curious.

dotnwat reviewed (commented):
generally looks good to me. seems like a merge conflict and some good questions in other reviews. will do a full review agian too.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/65f0160ce44e421cd1159820ba4afeedc8b71bec..2c74d2b6d9a0a3b9e6e7e2afec3bc5b5e99ee78f)
- Rebased against dev to resolve merge conflicts

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2c74d2b6d9a0a3b9e6e7e2afec3bc5b5e99ee78f..2b79e28d655191f48e85830224477cb7ddd2560c)
- Moved changes from `topics_frontend` into `security_frontend`
- Implemented Bens feedback that prevents raw cast from `std::error_code` to `cluster::errc`

dotnwat commented on src/v/kafka/server/handlers/describe_acls.cc:
```diff
@@ -88,7 +75,14 @@ ss::future<response_ptr> describe_acls_handler::handle(
+          filter, default_describe_acls_timeout);
+        if (result.error != cluster::errc::success) {
+            vlog(klog.debug, "Error describing ACLs: {}", result.error);
+            data.error_code = map_topic_error_code(result.error);
```
i'd consider making the timeout a bit longer like 5 seconds or something. and i wonder if we were to experience a timeout if we should just return what we have locally. mmm, i guess maybe not. btw, what do we return? shoudl we return not leader or something like that?

dotnwat commented on src/v/cluster/security_frontend.cc:
```diff
@@ -339,4 +339,98 @@ security_frontend::get_bootstrap_user_creds_from_env() {
+  model::node_id leader,
+  security::acl_binding_filter filter,
+  model::timeout_clock::duration timeout) {
+    vlog(clusterlog.trace, "Dispatching describe_acls to {}", leader);
```
can we have a ducktape test that asserts this line is present in the redpanda logs for some ACL test?

BenPope commented on src/v/cluster/security_frontend.cc:
```diff
@@ -339,4 +339,98 @@ security_frontend::get_bootstrap_user_creds_from_env() {
+              .operation = acl.operation(),
+              .permission_type = acl.permission(),
+            };
+            resource.acls.push_back(std::move(desc));
+        }
+        resources.push_back(std::move(resource));
```
The `emplace_back` suggestion was marked resolved without comment, so just wondered if it was forgotten.

BenPope commented on src/v/kafka/server/handlers/describe_acls.cc:
```diff
@@ -25,54 +27,39 @@
 template<>
 ss::future<response_ptr> describe_acls_handler::handle(
   request_context ctx, [[maybe_unused]] ss::smp_service_group ssg) {
+    static const auto default_describe_acls_timeout = 3s;
```
Can this default be overridden?
